### PR TITLE
Solvation plots

### PR DESCRIPTION
**Pull request by laurlee**
## Description
Provide a brief description of the PR's purpose here.

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ ] TODO 1

## Questions
- [ ] Question1

## Status
- [ ] Ready to go

### CODE REVIEW

File: solvation_analysis/tests/test_plotting.py
Comment (hmacdope) [inline solvation_analysis/tests/test_plotting.py:15]: In general there is no need to test that plotting works IMO. 
This is because these tests are not really very informative (they will always pass). 

As long as you have verified the plotting yourself I am happy with that but @orionarcher  may have a different opinion.

File: solvation_analysis/plotting.py
Comment (hmacdope) [inline solvation_analysis/plotting.py:41]: This feels a little bit unnecessary. I would just update these in the graph when required. :)

File: solvation_analysis/plotting.py
Comment (hmacdope) [inline solvation_analysis/plotting.py:109]: Reminder to fill these out :)

File: solvation_analysis/plotting.py
Comment (hmacdope) [inline solvation_analysis/plotting.py:270]: This should now be in branch as #59 has been merged.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Thanks for opening a new PR @laurlee!

Just a few comments. :) Are you planning to finish off the plotting functionality all in this PR?

File: solvation_analysis/tests/test_plotting.py
Comment (orionarcher) [reply solvation_analysis/tests/test_plotting.py:15]: I agree they aren't informative from the perspective of CI but I think they are informative for the developer. In my experience, visualization tests make it easier to use test-driven development and to iterate later on. If you are concerned about the CI time, we could comment out these lines:

```python
fig = plot_network_size_histogram(networking)
fig.show()
```

so that the tests are super cheap.

Comment (hmacdope) [reply solvation_analysis/tests/test_plotting.py:15]: Yeah I think commenting out fig.show is a good compromise. :)

Comment (hmacdope) [reply solvation_analysis/tests/test_plotting.py:15]: @orionarcher its your project so your call at the end of the day. :)

File: solvation_analysis/plotting.py
Comment (orionarcher) [reply solvation_analysis/plotting.py:41]: What do you think of the idea of a general templating function? I am thinking each function could return a fairly bare bones plot that could then be prettied up with some formatting function.

```python
def format_plot(fig)
    # format fig
    return fig

fig = plot_something(data)
pretty_fig = format_plot(fig)

```

This might be overly complicated, but we could also make it a decorator.

```python
def format_plot(plotting_func):
    def pretty_plotting_func(data):
        fig = plotting_func(data)
        # format fig
        return fig
    return pretty_plotting_func
```

we could then add the formatting decorator to our definition of each plot. This would allow solvation_analysis to standardize a particular plotting style (I have become quite opinionated on the subject) with maximum code reuse.

```python
@pretty_plot
def format_plot(data)
    # make fig
    return fig 
```

Comment (hmacdope) [reply solvation_analysis/plotting.py:41]: Sounds like a good idea to me!

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:235]: I'd recommend wrapping all of this logic into a single, well-named function. It should take the solvents from each electrolyte (`solvents_list` here) and keep_solvents as arguments and either throw an error or return keep_solvents.

You'll need to repeat this logic for every `compare_*` function, so best to package it up.

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:240]: If we are executing the `keep_solvents` parsing logic here then we don't need it as a kwarg for `compare_solvent_dicts`. We could put that logic in either place, which do you think would be better?

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:260]: I think we decided to deprecate the `ignore` kwarg?

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:59]: can we normalize the y-axis to "Fraction of All Networks"?

**Review by orionarcher — CHANGES_REQUESTED** [review]
Nice progress @laulee! A few comments, some of which we have already discussed.

File: solvation_analysis/tests/test_plotting.py
Comment (orbeckst) [inline solvation_analysis/tests/test_plotting.py:17]: I don't think you ever need to have `show()` in library code (and certainly not in the tests): either you can return the fig object and then the user can do what they want (including show) or your code plot to files and generates files (and then it slows you down a lot to run show).

File: solvation_analysis/tests/test_plotting.py
Comment (orbeckst) [inline solvation_analysis/tests/test_plotting.py:18]: I'd always assert something meaningful, e.g., that this is the correct instance, or you can dig into the figure object and check that it has the right labels, or that the data is what you expect it to be. But do a tiny bit more than flushing code paths.

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:198]: I'd replace "graphed in an order specified by the user" with something a bit more descriptive. Also probably a best practice to not refer to "the user" in the documentation itself. That's who is reading it!

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:195]: Let's think of a slightly more explicit name for coerce, maybe `coerce_solvent_names`

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:33]: I'd favor removing the `format_graph` function and just moving this to `plot_pairing`, `plot_coordination` etc. That will reduce the number of kwargs to `compare_solvent_dicts` and reduce the number of functions.

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:271]: is this an old implementation?

**Review by orionarcher — COMMENTED** [review]
Just a few small things. This is coming along nicely!

File: solvation_analysis/plotting.py
Comment (laurlee) [reply solvation_analysis/plotting.py:271]: Yes, I'll update it to reflect the current implementation when I next push the code.

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:216]: Let's move away calling anything a `solution` if it isn't a `Solution` object. If it's a `str` solution name, we should call it `solution_name`. Well chosen variable names make it far easier to understand the code!

There are a few other places where this needs to be fixed too.

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:130]: I know I originally suggested this name, but `coerce_solvent_names` is a bad variable name. We could change it to `new_solvent_name_dict` or something like that.

In Pycharm, you can use CMD + F6 to rename a variable.

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:132]: maybe `solvents_to_plot` instead of `keep_solvents`. We should also document what the behavior is if this is not set.

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:157]: If possible, let's try to avoid the `try except` loop.

Look into `Dict.get()` which returns the item if it exists and returns `None` if not. It's a very convenient function!

File: solvation_analysis/plotting.py
Comment (orionarcher) [inline solvation_analysis/plotting.py:289]: lets use `Solution.hasattr("residence")` to check if residence is present.

**Review by orionarcher — COMMENTED** [review]
Nice work @laurlee! I have a few comments mostly related to variable naming and readability. I think the implementation all looks good.

The main remaining challenge before merging this will be updating to `v0.1.4` which has a number of API changes.

**Pull request merged**
